src/BowlingScorer.js

```diff
@@ -0,0 +1,179 @@
+var BowlingScorer = function() {
+  this.total = 0;
+  this.runningTotals = [];
+  this.firstshot = null;
+  this.spare = false;
+  this.strike = false;
+  this.doubleStrike = false;
+  this.frameNo = 1;
+  this.semifinalShot = null;
+};
+
+BowlingScorer.prototype.addPins = function(pins,frameNo) {
+  this.total += pins;
+  if(frameNo > 9) {
+    frameNo=9
+    this.runningTotals[frameNo] += pins;
+  } else {
+    this.runningTotals[frameNo] = pins;
+  }
+};
+
+BowlingScorer.prototype.spareAdd = function(pins) {
+  this.addPins((pins+10),(this.frameNo-2));
+  this.spare = false
+};
+
+BowlingScorer.prototype.strikeAdd = function(pins) {
+  this.addPins((pins+10),(this.frameNo-2))
+  this.strike = false
+};
+
+BowlingScorer.prototype.doubleStrikeAdd = function(pins) {
+  this.addPins(pins+20,this.frameNo-3)
+  this.doubleStrike = false
+};
+
+BowlingScorer.prototype.bowl = function(pins) {
+  if(pins>=0&&pins<=10&&(pins%1===0)&&(this.firstshot)&&(this.frameNo<=10)) {
+    var frame = pins + this.firstshot;
+    if(frame<=10) {
+      if(this.strike) {
+        this.strikeAdd(frame);
+      }
+      if(frame===10) {
+        this.spare = true;
+        this.frameNo += 1
+        this.firstshot = null;
+        return pins;
+      }
+      this.addPins(frame,(this.frameNo-1));
+      this.frameNo += 1;
+      this.firstshot = null;
+      return pins;
+    }
+  } else if(pins>=0&&pins<=10&&(pins%1===0)&&(this.frameNo<=10)) {
+    if(this.spare) {
+      this.spareAdd(pins);
+    } else if(this.doubleStrike) {
+      this.doubleStrikeAdd(pins);
+    }
+
+    if(pins===10) {
+      if(this.strike) {
+        this.doubleStrike = true;
+        this.frameNo += 1;
+        return pins;
+      }
+      this.strike = true;
+      this.frameNo += 1;
+      return pins
+    }
+    return this.firstshot = pins;
+  } else if(pins>=0&&pins<=10&&(pins%1===0)&&(this.frameNo<=11)&&(this.spare||this.strike)) {
+    if(this.doubleStrike) {
+      this.doubleStrikeAdd(pins)
+    }
+
+    if(this.spare) {
+      this.spareAdd(pins);
+      this.frameNo += 2;
+    } else if(this.strike) {
+      if(pins===10) {
+        this.doubleStrike = true;
+        this.strikeAdd(pins)
+        this.frameNo += 1;
+      } else if(this.semifinalShot) {
+        if(pins+this.semifinalShot===10) { this.spare = true }
+        this.strikeAdd(pins+this.semifinalShot,(this.frameNo-2));
+        this.frameNo += 1;
+      }
+    }
+    return this.semifinalShot = pins
+  } else if(pins>=0&&pins<=10&&(pins%1===0)&&(this.frameNo<=12)&&(this.doubleStrike)) {
+    this.addPins(pins,this.frameNo-2);
+    this.frameNo += 1;
+    return pins;
+  }
+};
```

Comment:
Too many conditionals - refactor so that the code is reusable. Consider using an object for the rules, or an index for the strike / spare status.